### DISCUSSION TITLE
why can't numpy integers be supported directly?

### DISCUSSION THREAD

**Discussion by peterdhansen** [sqlalchemy/sqlalchemy#5948] [Usage Questions] [answered]
When you compare a column clause to a numpy integer, it works on the right but not on the left. I'd expect it to both ways or fail both ways. Simliar to #5846 

**To Reproduce**
```py
In [1]: from sqlalchemy import column
   ...: import numpy as np
   ...: column("hi") > np.int32(2)
   ...: np.int32(2) < column("hi")
```

**Error**

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-d000385c9e11> in <module>
      2 import numpy as np
      3 column("hi") > np.int32(2)
----> 4 np.int32(2) < column("hi")

site-packages\sqlalchemy\sql\elements.py in __bool__(self)
   3273             return self.operator(hash(self._orig[0]), hash(self._orig[1]))
   3274         else:
-> 3275             raise TypeError("Boolean value of this clause is not defined")
   3276
   3277     __nonzero__ = __bool__

TypeError: Boolean value of this clause is not defined
```

**Versions.**
 - OS: Windows 10
 - Python:  3.7.9
 - SQLAlchemy: 1.3.23

**Comment by zzzeek** [comment]
hi -

I haven't looked at this in some months but my recollection was that this is not possible w/ numpy because they implement their own comparison operators.  so when their value is on the left, theirs takes over.   #5846 was misnamed, in that Python does not actually have "reverse boolean" operators, that issue refers to bitwise operators (name is fixed).

background on the unsolvable numpy problem is at https://github.com/sqlalchemy/sqlalchemy/issues/5552#issuecomment-685033670 and an FAQ entry is at https://docs.sqlalchemy.org/en/13/faq/thirdparty.html#i-m-getting-errors-related-to-numpy-int64-numpy-bool-etc

**Comment by zzzeek** [comment, marked as answer]
from https://docs.sqlalchemy.org/en/13/faq/thirdparty.html#i-m-getting-errors-related-to-numpy-int64-numpy-bool-etc

The issue is due to the numpy.int64 datatype overriding the `__eq__()` method (and in this case other comparison methods) and enforcing that the return type of an expression is `numpy.True` or `numpy.False`, which breaks SQLAlchemy’s expression language behavior that expects to return ColumnElement expressions from Python equality comparisons:

```python
>>> import numpy
>>> from sqlalchemy import column, Integer
>>> print(column('x', Integer) == numpy.int64(10))  # works
x = :x_1
>>> print(numpy.int64(10) == column('x', Integer))  # breaks
False

```
These errors are both solved in the same way, which is that special numpy datatypes need to be replaced with regular Python values. Examples include applying the Python `int()` function to types like `numpy.int32` and `numpy.int64` and the Python `float()` function to `numpy.float32`:

```python
data = numpy.int64(10)

session.add(A(data=int(data)))

result = session.execute(
    select(A.data).where(int(data) == A.data)
)

session.commit()


```
